### ISSUE TITLE
When clearing canvas bitmap is broken

### ISSUE THREAD

**Issue by JosiahParry** [r-wasm/webr#335]
Here is an example doc that illustrates the issue. 

If i plot using the canvas I get a nice image
<img width="528" alt="image" src="https://github.com/r-wasm/webr/assets/13534508/9f2b556e-35b8-47c1-a6d0-a820da606e7a">

Then if i plot over it, it will overlay the image
<img width="504" alt="image" src="https://github.com/r-wasm/webr/assets/13534508/bb81ed60-42d1-4aa1-a67c-9f76c6eaeaa5">

So, naturally I choose to clear the canvas with `ctx.reset()`. However, when I do this, the subsequent drawing is incomplete? It seems the bit map gets damage here? 
<img width="531" alt="image" src="https://github.com/r-wasm/webr/assets/13534508/9e729ba5-7b49-44be-a4aa-174c5ff0c23e">

It's very likely i am doing something wrong here. If that's the case and you're willing to enlighten me, I'd be happy to try for a sentence or two in the docs as a PR :) 

```html
<html>
  <head>
    <title>WebR Test Console</title>
  </head>
  <body>
    <div>
      <canvas width="500" height="500" id="plot-canvas"></canvas>
    </div>
    <div>
      <pre><code id="out">Loading webR, please wait...</code></pre>
      <input spellcheck="false" autocomplete="off" id="input" type="text">
      <button onclick="globalThis.sendInput()" id="run">Run</button>
    </div>
    
    <script type="module">
      /* Create a webR console using the Console helper class */
      import { Console } from 'https://webr.r-wasm.org/latest/webr.mjs';
      const webRConsole = new Console({
        stdout: line => document.getElementById('out').append(line + '\n'),
        stderr: line => document.getElementById('out').append(line + '\n'),
        prompt: p => document.getElementById('out').append(p),
        canvasImage: ci => {
          var canvas = document.getElementById("plot-canvas")
          var ctx = canvas.getContext('2d');
          ctx.reset();
          ctx.drawImage(ci, 0, 0);
    },
      });
      webRConsole.run();
      
      /* Write to the webR console using the ``stdin()`` method */
      let input = document.getElementById('input');
      globalThis.sendInput = () => {
        webRConsole.stdin(input.value);
        document.getElementById('out').append(input.value + '\n');
        input.value = "";
      }
      
      /* Send input on Enter key */
      input.addEventListener(
        "keydown",
        (evt) => {if(evt.keyCode === 13) globalThis.sendInput()}
      );
    </script>
  </body>
</html>
```

**Comment by georgestagg** [comment]
Hi,

The issue is that a single plot from an R graphics device is made up of multiple drawing events, and so multiple `canvasImage` callbacks are fired for each plot. So, you shouldn't clear the canvas every time one of those events occurs, otherwise, you see partial plots consisting only of the final drawing session.

This is all based on the `mode` of the graphics device, and the behaviour is actually documented deep in the [R internals manual](https://cran.r-project.org/doc/manuals/r-release/R-ints.html#g_t_0027Mode_0027):

> [...] some screen devices ensure that drawing is flushed to the screen when called with mode = 0. It is tempting to use it for some sort of buffering, but note that ‘drawing’ is interpreted at quite a low level and a typical single figure will stop and start drawing many times.

The [`Console`](https://docs.r-wasm.org/webr/latest/api/js/classes/WebR.Console.html) class has a callback [`canvasNewPage`](https://docs.r-wasm.org/webr/latest/api/js/interfaces/WebR.ConsoleCallbacks.html#canvasnewpage) that fires when a new plot page is created. Your canvas should instead be cleared in that callback:

```javascript
  import { Console } from 'https://webr.r-wasm.org/latest/webr.mjs';

  var canvas = document.getElementById("plot-canvas")
  var ctx = canvas.getContext('2d');

  const webRConsole = new Console({
    stdout: line => document.getElementById('out').append(line + '\n'),
    stderr: line => document.getElementById('out').append(line + '\n'),
    prompt: p => document.getElementById('out').append(p),
    canvasImage: ci => ctx.drawImage(ci, 0, 0),
    canvasNewPage: () => ctx.reset(),
  });
  webRConsole.run();
```

I can see this is not well documented, so a PR along the lines you suggest is most welcome.

**Comment by SugarRayLua** [comment]
Hi.

I'm trying to follow the advice mentioned above and on the webR website re: using canvasNewPage. However, when I do so, I actually get the opposite effect: none of my images draw (it seems that either canvasNewPage is firing twice and erasing what I plot or somehow inhibiting canvasImage). Here is my sample below (only changed formatting slightly from online template; uncomment canvasNewPage() line to demonstrate problem):

```
<html>
  <head>
    <title>WebR Test Console</title>
  </head>
  <body>
    <div id="plot-output" style= "position: fixed; left: 55%; top: 10%">
      <canvas width="500" height="500" id="plot-canvas"></canvas>
    </div>
    <div>
      <pre><code id="out">Loading webR, please wait...</code></pre>
      <input spellcheck="false" autocomplete="off" id="input" type="text">
      <button onclick="globalThis.sendInput()" id="run">Run</button>
    </div>
    
    <script type="module">
      /* Create a webR console using the Console helper class */
      import { Console } from 'https://webr.r-wasm.org/latest/webr.mjs';

      var canvas = document.getElementById("plot-canvas")
      var ctx = canvas.getContext('2d');

      const webRConsole = new Console({
        stdout: line => document.getElementById('out').append(line + '\n'),
        stderr: line => document.getElementById('out').append(line + '\n'),
        prompt: p => document.getElementById('out').append(p),
        canvasImage: ci => ctx.drawImage(ci, 0, 0),
        //canvasNewPage: () => ctx.reset()
      });
      webRConsole.run();

      /* Set the default graphics device to be half the canvas element size */
      await webRConsole.stdin("options(device=webr::canvas(250, 250))");
      
      /* Write to the webR console using the ``stdin()`` method */
      let input = document.getElementById('input');
      globalThis.sendInput = () => {
        webRConsole.stdin(input.value);
        document.getElementById('out').append(input.value + '\n');
        input.value = "";
      }
      
      /* Send input on Enter key */
      input.addEventListener(
        "keydown",
        (evt) => {if(evt.keyCode === 13) globalThis.sendInput()}
      );
    </script>
  </body>
</html>
```

device tested on: iPad OS 16.7.2 mobile browsers: mobile Safari, mobile Edge, mobile iCabMobile

Thanks!

**Comment by SugarRayLua** [comment]
I apologize; I figured out the problem: the example above and the example on the website use the canvas context method:

"reset()"

I read that the reset() method is fairly new for some browsers including mobile Safari browser. I was using an older mobile Safari browser that didn't incorporate that method. I realized as a workaround, I could use the older canvas context method:

"clearRect(0,0,w,h)"

and pass in the width and height attributes of my canvas which also clears the screen and works. 

Fyi in case others have the same problem
